optuna/storages/in_memory.py

```diff
@@ -53,108 +51,131 @@ def __setstate__(self, state):
+            for trial in self._studies[study_id].trials:
+                del self._trial_id_to_study_id_and_number[trial._trial_id]
+            study_name = self._studies[study_id].name
+            del self._study_name_to_id[study_name]
+            del self._studies[study_id]
 
     def set_study_direction(self, study_id, direction):
         # type: (int, StudyDirection) -> None
 
+        self._check_study_id(study_id)
+
         with self._lock:
-            if self.direction != StudyDirection.NOT_SET and self.direction != direction:
+            if (
+                self._studies[study_id].direction != StudyDirection.NOT_SET
+                and self._studies[study_id].direction != direction
+            ):
                 raise ValueError(
                     "Cannot overwrite study direction from {} to {}.".format(
-                        self.direction, direction
+                        self._studies[study_id].direction, direction
                     )
                 )
-            self.direction = direction
+            self._studies[study_id].direction = direction
 
     def set_study_user_attr(self, study_id, key, value):
         # type: (int, str, Any) -> None
 
+        self._check_study_id(study_id)
+
         with self._lock:
-            self.study_user_attrs[key] = value
+            self._studies[study_id].user_attrs[key] = value
 
     def set_study_system_attr(self, study_id, key, value):
         # type: (int, str, Any) -> None
 
+        self._check_study_id(study_id)
+
         with self._lock:
-            self.study_system_attrs[key] = value
+            self._studies[study_id].system_attrs[key] = value
 
     def get_study_id_from_name(self, study_name):
         # type: (str) -> int
 
-        if study_name != self.study_name:
+        if study_name not in self._study_name_to_id:
             raise ValueError("No such study {}.".format(study_name))
 
-        return IN_MEMORY_STORAGE_STUDY_ID
+        return self._study_name_to_id[study_name]
 
     def get_study_id_from_trial_id(self, trial_id):
         # type: (int) -> int
 
-        return IN_MEMORY_STORAGE_STUDY_ID
+        self._check_trial_id(trial_id)
+
+        return self._trial_id_to_study_id_and_number[trial_id][0]
 
     def get_study_name_from_id(self, study_id):
         # type: (int) -> str
 
         self._check_study_id(study_id)
-        return self.study_name
+        return self._studies[study_id].name
 
     def get_study_direction(self, study_id):
         # type: (int) -> StudyDirection
 
-        return self.direction
+        self._check_study_id(study_id)
+        return self._studies[study_id].direction
 
     def get_study_user_attrs(self, study_id):
         # type: (int) -> Dict[str, Any]
 
+        self._check_study_id(study_id)
         with self._lock:
-            return copy.deepcopy(self.study_user_attrs)
+            return self._studies[study_id].user_attrs
```

Comment:
Yes, it can be addressed into another PR.
However, the current implementation of this PR relies on this copying behavior of get_trial, and modifying the line requires a large change of codes.

I'll create another PR to change the copying behavior and I'd like to merge it before this PR.

Reply:
After examining the code, I decided not to create separate PR and keep the original copying behavior in this PR.